plugin-server/src/main/ingestion-queues/batch-processing/each-batch.ts

```diff
@@ -72,3 +78,92 @@ export async function eachBatch(
         transaction.finish()
     }
 }
+
+export async function eachBatchParallel(
+    /**
+     * Using the provided groupIntoBatches function, split the incoming batch into micro-batches
+     * that are executed **in parallel**, committing offsets only at the end.
+     * Events within a single micro-batch are processed **sequentially**.
+     */
+    { batch, resolveOffset, heartbeat, commitOffsetsIfNecessary, isRunning, isStale }: EachBatchPayload,
+    queue: IngestionConsumer,
+    eachMessage: (message: PipelineEvent, queue: IngestionConsumer) => Promise<void>,
+    groupIntoBatches: (messages: KafkaMessage[]) => PipelineEvent[][],
+    key: string
+): Promise<void> {
+    const batchStartTimer = new Date()
+    const loggingKey = `each_batch_${key}`
+
+    const transaction = Sentry.startTransaction(
+        { name: `eachBatchParallel(${eachMessage.name})` },
+        { topic: queue.topic }
+    )
+
+    try {
+        /**
+         * Micro-batches should be executed from biggest to smallest to enable the best concurrency.
+         * We're sorting with biggest last and pop()ing. Ideally, we'd use a priority queue by length
+         * and a separate array for single messages, but let's look at profiles before optimizing.
+         */
+        const messageBatches = groupIntoBatches(batch.messages)
+        messageBatches.sort((a, b) => a.length - b.length)
+
+        queue.pluginsServer.statsd?.histogram('ingest_event_batching.input_length', batch.messages.length, { key: key })
+        queue.pluginsServer.statsd?.histogram('ingest_event_batching.batch_count', messageBatches.length, { key: key })
+
+        async function processMicroBatches(batches: PipelineEvent[][]): Promise<void> {
+            let currentBatch
+            let processedBatches = 0
+            while ((currentBatch = batches.pop()) !== undefined) {
+                const batchSpan = transaction.startChild({
+                    op: 'messageBatch',
+                    data: { batchLength: currentBatch.length },
+                })
+                for (const message of currentBatch) {
+                    await Promise.all([eachMessage(message, queue), heartbeat()])
```

Comment:
I'm guessing `heartbeat` is like commitifrequired or something i.e .it doesn't always do a heartbeat to Kafka?

Reply:
Indeed, [the docs](https://kafka.js.org/docs/consuming#a-name-each-batch-a-eachbatch) state: `if you invoke heartbeat() sooner than heartbeatInterval it will be ignored.`